Build consolidation

metlos opened the pull request:
### What does this PR do?
Build consolidation:
    
* Makefile and deployment generation now use environment variables prefixed
with DWCO_ for parts that might conflict with devworkspace-operator
* Makefile no longer uses `install` and `uninstall` targets because those
  made no sense anymore.
* `make run` runs the operator locally with the perms of the generated
   service account
* `make run_as_current_user` to run as the user in the current kubeconfig
  context
* `make prepare` only installs rbac, crd, ..., but NOT the operator deployment
  I.e. it makes the cluster ready for running the operator locally
* `make deploy` deploys everything into the cluster, including a Deployment
  with the configured image
* `make undeploy` removes everything dwco related from the cluster,
  including any workspaces, routings, etc.

### What issues does this PR fix or reference?
eclipse/che#19286

metlos committed:
Use a prefix for the env var names influencing the build. This is so that
users that also build devworkspace-operator can safely export their
env vars without fearing the clash between the respective builds.

metlos committed:
Build consolidation:

* Makefile and deployment generation now use environment variables prefixed
with DWCO_ for parts that might conflict with devworkspace-operator
* Makefile no longer uses `install` and `uninstall` targets because those
  made no sense anymore.
* `make run` runs the operator locally with the perms of the generated
   service account
* `make run_as_current_user` to run as the user in the current kubeconfig
  context
* `make prepare` only installs rbac, crd, ..., but NOT the operator deployment
  I.e. it makes the cluster ready for running the operator locally
* `make deploy` deploys everything into the cluster, including a Deployment
  with the configured image
* `make undeploy` removes everything dwco related from the cluster,
  including any workspaces, routings, etc.

metlos commented:
Manual testing succeeded. Devworkspace deployment seems to work and `DevworkspaceRouting` objects are correctly reconciled.

metlos committed:
Rename deploy/undeploy to install/uninstall to be inline with
devworkspace-operator.

skabashnyuk commented on Makefile:
```diff
@@ -27,38 +33,61 @@ else
-run: generate fmt vet manifests
+### prepare: Prepares the cluster for running the operator - deploys CRDs, rbac and the service account
+prepare: _mk_temp manifests
+	$(eval PREPAPRE_OUTPUT_DIR := $(shell mktemp -p $(TEMP_DIR) -d))
```
maybe it's better to move the long command to a separate bash script?
Motivation - bash script format is simpler than Makefile format.

metlos replied:
That script would need to rely on 4 parameters (or env vars) and would be 5 lines.. That's why I opted for this 5-lines of make target. I will add some comments to explain what's going on there.

skabashnyuk commented on Makefile:
```diff
@@ -27,38 +33,61 @@ else
+	@rm -Rf $(TEMP_DIR)
 
 ### test: Run tests
 test: generate fmt vet manifests
```
Should we link this to https://github.com/che-incubator/devworkspace-che-operator/blob/main/.github/workflows/test.yml#L64 ?

metlos replied:
we could

skabashnyuk replied:
make sense to do it with this pr or we can postpone that?

metlos replied:
I did it in this PR. See the later commits.

metlos committed:
Use the test target of the Makefile in github actions.
Some explanatory comments in the Makefile.

metlos committed:
Use `ci` tag for the main branch builds. Also tag the builds with a sha of
the commit so that the code from which the image was built can be easily
identified.

metlos committed:
Merge remote-tracking branch 'upstream/main' into build-consolidation

metlos committed:
ah, the deployment files...
